File: 2000290310115/Week1/Linkedlist.cpp
#include <bits/stdc++.h>
using namespace std;
//node creation
class Node {
public:
    int data;
    Node* next;

    Node(int data) {
        this->data = data;
        this->next = NULL;
    }
};
class LinkedList {
private:
    Node* head;
public:
    LinkedList() {
        head = NULL;
    }
//Insertion
    void insertAtHead(int data) {
        Node* newNode = new Node(data);
        newNode->next = head;
        head = newNode;
    }

    void insertAtTail(int data) {
        Node* newNode = new Node(data);
        if (head == NULL) {
            head = newNode;
        } else {
            Node* current = head;
            while (current->next != NULL) {
                current = current->next;
            }
            current->next = newNode;
        }
    }

    void insertAtPosition(int data, int position) {
        if (position <= 0) {
            std::cout << "Invalid position." << std::endl;
            return;
        }
        if (position == 1 || head == NULL) {
            insertAtHead(data);
        } else {
            Node* newNode = new Node(data);
            Node* current = head;
            int count = 1;
            while (count < position - 1 && current != NULL) {
                current = current->next;
                count++;
            }
            if (current == NULL) {
                std::cout << "Invalid position." << std::endl;
                return;
            }
            newNode->next = current->next;
            current->next = newNode;
        }
    }
//Deletion
    void deleteAtPosition(int position) {
        if (position <= 0 || head == NULL) {
            std::cout << "Invalid position." << std::endl;
            return;
        }
        if (position == 1) {
            Node* temp = head;
            head = head->next;
            delete temp;
        } else {
            Node* current = head;
            Node* prev = NULL;
            int count = 1;
            while (count < position && current != NULL) {
                prev = current;
                current = current->next;
                count++;
            }
            if (current == NULL) {
                std::cout << "Invalid position." << std::endl;
                return;
            }
            prev->next = current->next;
            delete current;
        }
    }
//Searching
    bool search(int value) {
        Node* current = head;
        while (current != NULL) {
            if (current->data == value) {
                return true;
            }
            current = current->next;
        }
        return false;
    }
//Display
    void print() {
        Node* current = head;
        while (current != NULL) {
            std::cout << current->data << " ";
            current = current->next;
        }
        cout << endl;
    }
};

int main() {
    LinkedList myLst;
    myLst.insertAtHead(1);
    myLst.insertAtHead(2);

    myLst.insertAtTail(4);
    myLst.insertAtTail(3);

    myLst.insertAtPosition(10, 2);
    myLst.insertAtPosition(20, 3);
  
    myLst.print(); 

    myLst.deleteAtPosition(4);
    myLst.deleteAtPosition(1);

    myLst.print();  

    cout << myLst.search(20) << endl; 
    cout << myLst.search(1) <<endl; 

    return 0;
}

File: 2000290120194/WEEK1/linkedlist.java
#LINKEDLIST IMPLEMENTATION


class Node {
    int data;
    Node next;

    public Node(int data) {
        this.data = data;
        this.next = null;
    }
}

class LinkedList {
    Node head;

    public LinkedList() {
        this.head = null;
    }

    // Insertion at the end of the list
    public void insert(int data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
        } else {
            Node current = head;
            while (current.next != null) {
                current = current.next;
            }
            current.next = newNode;
        }
    }

    // Insertion at the beginning of the list
    public void insertAtBeginning(int data) {
        Node newNode = new Node(data);
        newNode.next = head;
        head = newNode;
    }

    // Insertion at a specific position in the list
    public void insertAtPosition(int data, int position) {
        if (position < 0) {
            throw new IllegalArgumentException("Position cannot be negative.");
        }

        if (position == 0) {
            insertAtBeginning(data);
            return;
        }

        Node newNode = new Node(data);
        Node current = head;
        for (int i = 0; i < position - 1; i++) {
            if (current == null) {
                throw new IndexOutOfBoundsException("Position exceeds the size of the list.");
            }
            current = current.next;
        }

        newNode.next = current.next;
        current.next = newNode;
    }

    // Deletion of a node with given data value
    public void delete(int data) {
        if (head == null) {
            return;
        }

        if (head.data == data) {
            head = head.next;
            return;
        }

        Node current = head;
        while (current.next != null) {
            if (current.next.data == data) {
                current.next = current.next.next;
                return;
            }
            current = current.next;
        }
    }

    // Display the linked list
    public void display() {
        Node current = head;
        while (current != null) {
            System.out.print(current.data + " ");
            current = current.next;
        }
        System.out.println();
    }
}

public class Main {
    public static void main(String[] args) {
        LinkedList linkedList = new LinkedList();

        linkedList.insert(1);
        linkedList.insert(2);
        linkedList.insert(3);

        linkedList.display(); // Output: 1 2 3

        linkedList.insertAtBeginning(0);
        linkedList.display(); // Output: 0 1 2 3

        linkedList.insertAtPosition(4, 4); // Inserting 4 at index 4 (which is out of range)
        // Output: IndexOutOfBoundsException: Position exceeds the size of the list.

        linkedList.insertAtPosition(4, 3); // Inserting 4 at index 3
        linkedList.display(); // Output: 0 1 2 4 3

        linkedList.delete(2);
        linkedList.display(); // Output: 0 1 4 3
    }
}

